src/persist/src/file.rs

```diff
@@ -0,0 +1,197 @@
+        self.buf.clear();
+        for ((key, val), ts, diff) in updates.iter() {
+            let key_bytes = key.as_bytes();
+            let key_bytes_len = key_bytes.len() as u64;
+            let val_bytes = val.as_bytes();
+            let val_bytes_len = val_bytes.len() as u64;
+
+            self.buf.extend(&key_bytes_len.to_le_bytes());
+            self.buf.extend(key_bytes);
+            self.buf.extend(&val_bytes_len.to_le_bytes());
+            self.buf.extend(val_bytes);
+            self.buf.extend(&ts.to_le_bytes());
+            self.buf.extend(&(*diff as i64).to_le_bytes());
+        }
+
+        self.buf.shrink_to_fit();
+
+        let mut guard = self.dataz.lock()?;
+        guard.write_all(&self.buf)?;
+        // TODO: we will eventually want to change the interface to let us batch up
+        // calls to `sync_all`.
+        guard.sync_all()?;
+        Ok(())
+    }
+}
+
+impl Meta for FileStream {
+    type Snapshot = MemSnapshot;
+
+    fn snapshot(&self) -> Result<Self::Snapshot, Error> {
+        let mut bytes = Vec::new();
+        let mut dataz = Vec::new();
+
+        {
+            let mut guard = self.dataz.lock()?;
+            guard
+                .seek(SeekFrom::Start(0))
+                .expect("failed to seek to beginning");
+            guard.read_to_end(&mut bytes).expect("failed to read file");
+        }
+
+        let mut bytes = &bytes[..];
+        let mut buf = [0u8; 8];
+        while !bytes.is_empty() {
+            // Decode key
+            bytes.read_exact(&mut buf)?;
+            let key_len = u64::from_le_bytes(buf);
+            let mut key: Vec<u8> = vec![0u8; key_len.try_into().unwrap()];
+            bytes.read_exact(&mut key[..])?;
+            let key = String::from_utf8(key).expect("failed to parse key");
+
+            // Decode val
+            bytes.read_exact(&mut buf)?;
+            let val_len = u64::from_le_bytes(buf);
+            let mut val = vec![0u8; val_len.try_into().unwrap()];
+            bytes.read_exact(&mut val[..])?;
+            let val = String::from_utf8(val).expect("failed to pass val");
+
+            // Decode time
+            bytes.read_exact(&mut buf)?;
+            let ts = u64::from_le_bytes(buf);
+
+            // Decode diff
+            bytes.read_exact(&mut buf)?;
+            let diff = i64::from_le_bytes(buf);
+
+            dataz.push(((key, val), ts, diff as isize));
+        }
+        Ok(MemSnapshot { dataz })
+    }
+
+    fn allow_compaction(&mut self, _ts: u64) -> Result<(), Error> {
+        // No-op for now.
+        Ok(())
+    }
+}
+
+/// Implementation of [Persister]
+pub struct FilePersister {
+    base_path: PathBuf,
+    registered: HashSet<Id>,
+    dataz: HashMap<Id, FileStream>,
+}
+
+impl FilePersister {
+    /// Constructs a new, empty SimplePersister.
+    pub fn new<P: AsRef<Path>>(base_path: P) -> Self {
+        std::fs::create_dir_all(&base_path).expect("failed to create FilePersister base path");
```

Comment:
I think it's idiomatic to import `std::fs` and then use `fs::create_dir_all` here

also return a result instead of expect